src/interpolate.f90

```diff
@@ -0,0 +1,177 @@
+module interpolate
+    implicit none
+
+    integer, parameter :: MAX_ORDER = 5
+
+    type :: SplineData1D
+        integer :: order
+        integer :: num_points
+        logical :: periodic
+        real(8) :: h_step
+        real(8), dimension(:,:), allocatable :: coeff
+    end type SplineData1D
+
+    type :: SplineData2D
+        integer :: order(2)
+        integer :: num_points(2)
+        logical :: periodic(2)
+        real(8) :: h_step(2)
+        real(8), dimension(:,:,:,:), allocatable :: coeff
+    end type SplineData2D
+
+    type :: SplineData3D
+        integer :: order(3)
+        integer :: num_points(3)
+        logical :: periodic(3)
+        real(8) :: h_step(3)
+        real(8), dimension(:,:,:,:,:,:), allocatable :: coeff
+    end type SplineData3D
+
+contains
+
+    subroutine construct_splines_1d(x, y, order, periodic, spl)
+        real(8), intent(in) :: x(:), y(:)
+        integer, intent(in) :: order
+        logical, intent(in) :: periodic
+
+        type(SplineData1D), intent(out) :: spl
+
+        spl%order = order
+        spl%periodic = periodic
+        spl%num_points = size(x)
+        spl%h_step = x(2) - x(1)
+
+        allocate(spl%coeff(0:order, spl%num_points))
+        spl%coeff(0,:) = y
+
+        if (periodic) then
+            call spl_per(spl%order, spl%num_points, spl%h_step, spl%coeff)
+        else
+            call spl_reg(spl%order, spl%num_points, spl%h_step, spl%coeff)
+        endif
+    end subroutine construct_splines_1d
+
+
+    subroutine destroy_splines_1d(spl)
+        type(SplineData1D), intent(inout) :: spl
+
+        deallocate(spl%coeff)
+    end subroutine destroy_splines_1d
+
+
+    subroutine evaluate_splines_1d(x, spl, y)
+        real(8), intent(in) :: x
+        type(SplineData1D), intent(in) :: spl
+        real(8), intent(out) :: y
+
+        real(8) :: x_norm, x_local, local_coeff(0:MAX_ORDER)
+        integer :: interval_index, k_power
+
+        x_norm = x/spl%h_step
+        interval_index = max(0, min(spl%num_points-1, int(x_norm)))
+        x_local = (x_norm - dble(interval_index))*spl%h_step  ! Distance to grid point
+
+        local_coeff(0:spl%order) = spl%coeff(:, interval_index+1)
+
+        ! Start with largest power and then multiply recursively
+        y = local_coeff(spl%order)
+        do k_power = spl%order, 0, -1
+            y = local_coeff(k_power) + x_local*y
+        enddo
+    end subroutine evaluate_splines_1d
+
+
+    subroutine construct_splines_2d(x1, x2, y, order, periodic, spl)
+        real(8), intent(in) :: x1(:), x2(:), y(:,:)
+        integer, intent(in) :: order(2)
+        logical, intent(in) :: periodic(2)
```

Comment:
I suggest to use `order(:)` and `periodic(:)`, because

1. `size(order)` and `size(periodic)` would at least yield the correct result if used, and
2. with bounds checking enabled, the Fortran runtime would print an error message including the size mismatch and not just continue silently (or possibly with a segfault).

Size mismatch for explicit-shape arrays is only caught by the compiler if the actual argument is of fixed size, i.e. not allocatable.